homework/index.js

```diff
@@ -30,17 +32,152 @@
+          class: 'contributors-title',
+        });
+        if (!contributors.length) {
+          contributorsTitle.textContent = 'No Contributor So Far';
+          return;
+        }
+        const contributorsList = createAndAppend('ul', rightSide, {
+          class: 'contributors-list',
+        });
+        contributors.forEach(contributor => {
+          const listItem = createAndAppend('li', contributorsList, {
+            class: 'contributor',
+            tabindex: 0,
+            'aria-label': contributor.login,
+          });
+          createAndAppend('img', listItem, {
+            class: 'contributor-avatar',
+            src: contributor.avatar_url,
+          });
+          createAndAppend('span', listItem, {
+            class: 'contributor-name',
+            text: contributor.login,
+          });
+          createAndAppend('span', listItem, {
+            class: 'contribution-count',
+            text: contributor.contributions,
+          });
+          listItem.addEventListener('click', () => {
+            // Go to new page on click
+            window.open(contributor.html_url, '_blank');
+          });
+          listItem.addEventListener('keyup', event => {
+            // Open new page with contributor url when pressed Enter on focused contributor
+            if (event.key === 'Enter') window.open(contributor.html_url, '_blank');
+          });
+        });
+      })
+      .catch(error => renderError(error));
+  }
+
+  function initializePage(repositories) {
+    const select = document.getElementById('select');
+    const wrapper = document.getElementById('wrapper');
+    repositories.sort((a, b) => a.name.localeCompare(b.name));
+    repositories.forEach((repository, index) => {
+      createAndAppend('option', select, {
+        value: index,
+        text: repository.name,
+      });
+    });
+    createAndAppend('section', wrapper, {
+      class: 'left-side',
+      id: 'left-side',
+    });
+    renderRepoInfo(repositories[select.value]);
+    createAndAppend('section', wrapper, {
+      class: 'right-side',
+      id: 'right-side',
+    });
+    select.addEventListener('change', () => {
+      renderRepoInfo(repositories[select.value]);
+      fetchAndRender(repositories[select.value]);
+    });
+    return repositories[select.value];
+  }
+
+  function renderPage(url) {
+    const root = document.getElementById('root');
+    const header = createAndAppend('header', root, {
+      class: 'header',
+    });
+    createAndAppend('h1', header, {
+      class: 'header-title',
+      text: 'Hack Your Future Repositories',
+    });
+    createAndAppend('main', root, {
+      id: 'wrapper',
+    });
+    createAndAppend('select', header, {
+      id: 'select',
+      'aria-label': 'HYF Repositories',
+    });
+
+    fetchJSON(url)
+      .then(repositories => initializePage(repositories))
+      .then(repository => fetchAndRender(repository))
+      .catch(error => renderError(error));
```

Comment:
`initiazePage` does not return a promise by itself. But this still works because the first `.then()` return a promise that is immediately resolved to the return value of `initializePage`. Alternatively, you could code it likes this:

```js
fetchJSON(url)
  .then(repositories => {
    const repository = initializePage(repositories);
    return fetchAndRender(repository);
  })
  .catch(error => renderError(error));
```

On the plus side, your implementation reads nicely. On the downside, it implies something that does not reflect reality.

Reply:
Changed the code following your advice. Actually i trusted the `then`'s promise wrapping mechanism if the result is not a promise.